doc/tutorial/riscv-virtual-memory.rst

```diff
@@ -0,0 +1,413 @@
+
+    .. collapse:: Solution
+
+        The translation is found in the TLB.
+        Clear the TLB by executing ``cpu0 tlbflush`` (more on this later) and try again.
+        
+        How does the translation differ now?
+
+        .. collapse:: Solution 2
+
+            Only one level of the pagetable is used.
+            This is because the code is mapped using a megapage.
+
+In addition to ``tr`` MSIM also supports the ``str`` command.
+Similarly to ``sptd``, you specify the used pagetable by its physical address.
+Also, ``str`` completely ignores the TLB.
+
+.. code:: msim
+
+    [msim] cpu0 str 0xA0000000 0xB0000000
+    VPN[1]: 0x2c0 VPN[0]: 0x000 page offset: 0x000
+    PTE1: [ PPN: 0x0a0001 RSW: 00 ---- ---V ]
+      This entry ^ physical address: 0x0a0000b00 = 0x0a0000000 + 0x2c0 * 4
+    PTE2: [ PPN: 0x0c0000 RSW: 00 ---- XWRV ]
+      This entry ^ physical address: 0x0a0001000 = 0x0a0001000 + 0x000 * 4
+
+    OK: 0xb0000000 => 0x0c0000000
+
+What even is the TLB?
+---------------------
+
+.. quiz:: 
+
+    You know, that is a good question. What even is the TLB?
+
+    .. collapse:: Hint
+
+        TLB stands for **T**\ ranslation **L**\ ookaside **B**\ uffer.
+
+    .. collapse:: Solution
+
+        TLB is a cache used to store virtual translation results.
+        It works on the level of pages (either 4 KiB or 4 MiB megapages).
+        
+        If we were to translate ``0x12345000 => 0x6789A000`` using a pagetable
+        (and thus reading twice from memory), we cache that the ``0x12345`` VPN is mapped
+        to the ``0x6789A`` PPN. Let's say we want to translate the address ``0x123450F0`` next.
+        We start by looking into the TLB and notice, that we have an entry for its VPN.
+        We can translate this address without looking inside of the pagetable. 
+        We do so and translate it to ``0x6789A0F0``.
+
+        The TLB entries are added automatically to a finite TLB, if there is not a free space for the new
+        entry, the **L**\ east **R**\ ecently **U**\ sed entry is evicted.
+        The ``sfence.vma`` instruction serves for manual eviction.
+        It can either clear the whole TLB or you can use it to evict based on the ASID, virtual address or both.
+
+The size of the RISC-V TLB is configurable in MSIM (using the ``tlbresize`` command),
+but using the default count of 48 entries should be reasonable for most applications.
+
+When translating an address the TLB is first searched for an entry
+which maps the given virtual address and which is either global
+or has the currently active ASID. Thus if you intend to use the same ASID
+for different address spaces, you have to carefully flush the TLB.
+It's always safe to flush the cache, but in the real world,
+you would try to flush out only the problematic entries, so that the TLB
+is used as much as possible and the CPU does not have to wait for slow
+memory reads.
```

Comment:
I would try to briefly recap why flushing is safe, something along the following lines might work? Thus if we intend to use the same ASID for two different address spaces, we would need to flush the TLB. Flushing TLB is always safe (TLB is a performance optimization, after all) but in the real world we should always try to flush only the conflicting entries, keeping the rest of TLB intact.